wallet/service.go

```diff
@@ -0,0 +1,177 @@
+	ctx context.Context,
+	walletID uuid.UUID,
+	transaction string,
+	destination string,
+) (*wallet.TransactionInfo, error) {
+	info, err := service.Datastore.GetWallet(walletID)
+	if err != nil {
+		return nil, errorutils.Wrap(err, "error getting wallet")
+	}
+	return service.SubmitCommitableAnonCardTransaction(ctx, info, transaction, destination, true)
+}
+
+// GetWallet - get a wallet by id
+func (service *Service) GetWallet(ID uuid.UUID) (*walletutils.Info, error) {
+	return service.Datastore.GetWallet(ID)
+}
+
+// SubmitCommitableAnonCardTransaction submits a transaction
+func (service *Service) SubmitCommitableAnonCardTransaction(
+	ctx context.Context,
+	info *walletutils.Info,
+	transaction string,
+	destination string,
+	confirm bool,
+) (*wallet.TransactionInfo, error) {
+	providerWallet, err := provider.GetWallet(ctx, *info)
+	if err != nil {
+		return nil, err
+	}
+	anonCard, ok := providerWallet.(*uphold.Wallet)
+	if !ok {
+		return nil, errors.New("Only uphold wallets are supported")
+	}
+
+	// FIXME needs to require the idempotency key
+	_, err = anonCard.VerifyAnonCardTransaction(transaction, destination)
+	if err != nil {
+		return nil, err
+	}
+
+	// Submit and confirm since we are requiring the idempotency key
+	return anonCard.SubmitTransaction(transaction, confirm)
+}
+
+// LinkWallet links a wallet and transfers funds to newly linked wallet
+func (service *Service) LinkWallet(
+	ctx context.Context,
+	wallet uphold.Wallet,
+	transaction string,
+	anonymousAddress *uuid.UUID,
+) error {
+	// do not confirm this transaction yet
+	info := wallet.GetWalletInfo()
+
+	var (
+		userID          string
+		depositProvider string
+		probi           decimal.Decimal
+	)
+
+	// verify that the user is kyc from uphold. (for all wallet provider cases)
+	if uID, ok, err := wallet.IsUserKYC(ctx, transaction); err != nil {
+		// there was an error
+		return handlers.WrapError(err,
+			"wallet could not be kyc checked",
+			http.StatusInternalServerError,
+		)
+	} else if !ok {
+		// fail
+		return &handlers.AppError{
+			Cause: errors.New("user kyc did not pass"),
+			Code:  http.StatusForbidden,
+		}
+	} else {
+		userID = uID
+	}
+
+	if wallet.Provider == "uphold" {
+		tx, err := service.SubmitCommitableAnonCardTransaction(
+			ctx,
+			&info,
+			transaction,
+			"",
+			false,
+		)
+		if err != nil {
+			return handlers.WrapError(err, "unable to verify transaction", http.StatusBadRequest)
+		}
+		if tx.UserID == "" {
+			err := errors.New("user id not provided")
+			return handlers.WrapError(err, "unable to link wallet", http.StatusBadRequest)
+		}
+		userID = tx.UserID
+		probi = tx.Probi
+	} else {
+		tx, err := wallet.VerifyTransaction(transaction)
+		if err != nil {
+			return handlers.WrapError(errors.New("failed to verify transaction"), "failed to verify transaction", http.StatusForbidden)
+		}
+		// get the original transaction probi amount
+		probi = tx.Probi
+		info.ProviderID = tx.Destination
```

Comment:
I believe it is required since all this logic was moved into the service.  We need to set the provider id on the wallet info to the tx.Destination in the case where this is user owned deposit account wallet.

Reply:
https://github.com/brave-intl/bat-go/pull/314/commits/9b6f3c0abbef3ac097cd9c30ee3c6d426dff5a53